src/main/src/main/java/org/geoserver/security/GeoServerSecurityManager.java

```diff
@@ -326,11 +351,25 @@ public void onApplicationEvent(ApplicationEvent event) {
 
             // migrate from old security config
             try {
-                boolean migratedFrom21 = migrateFrom21();
-                removeErroneousAccessDeniedPage();
-                migrateFrom22(migratedFrom21);
-                migrateFrom23();
-                migrateFrom24();                
+                Version securityVersion = getSecurityVersion();
```

Comment:
In order to avoid running a series of expensive security dir migrations I've added a version property file, that the code uses to figure out the starting version, from there it will do the migrations required and upgrade the version number. Blindly doing the 2.3 -> 2.4 and 2.4 -> 2.5 migratinos was the number one performance offender